packages/atomic/src/components/result-lists/result-list-common.tsx

```diff
@@ -0,0 +1,317 @@
+  ResultListProps,
+  EcommerceDefaultFieldsToInclude,
+} from '@coveo/headless';
+import {identity} from 'lodash';
+import {Bindings} from '../../utils/initialization-utils';
+import {
+  ResultDisplayDensity,
+  ResultDisplayImageSize,
+  ResultDisplayLayout,
+  getResultDisplayClasses,
+} from '../atomic-result/atomic-result-display-options';
+import {TemplateContent} from '../result-templates/result-template-common';
+import {TableDisplayResultsPlaceholder} from './table-display-results-placeholder';
+import {TableDisplayResults} from './table-display-results';
+import {ListDisplayResults} from './list-display-results';
+import {GridDisplayResults} from './grid-display-results';
+import {GridDisplayResultsPlaceholder} from './grid-display-results-placeholder';
+import {ListDisplayResultsPlaceholder} from './list-display-results-placeholder';
+import {once} from '../../utils/utils';
+import {updateBreakpoints} from '../../utils/replace-breakpoint';
+export interface AtomicResultListBaseComponent
+  extends Omit<ResultsProps, 'classes'> {
+  templateHasError: boolean;
+  listWrapperRef?: HTMLDivElement;
+  resultsPerPageState: ResultsPerPageState;
+
+  resultListCommon: ResultListCommon;
+
+  resultList: FoldedResultList | ResultList;
+}
+interface DisplayOptions {
+  density: ResultDisplayDensity;
+  imageSize?: ResultDisplayImageSize;
+  image?: ResultDisplayImageSize;
+  display?: ResultDisplayLayout;
+}
+export interface ResultPlaceholderProps extends Omit<DisplayOptions, 'image'> {
+  resultsPerPageState: ResultsPerPageState;
+}
+export interface ResultsProps extends DisplayOptions {
+  resultListState: FoldedResultListState | ResultListState;
+  resultListCommon: ResultListCommon;
+  bindings: Bindings;
+  host: HTMLElement;
+  getContentOfResultTemplate(
+    result: Result | FoldedResult
+  ): HTMLElement | DocumentFragment;
+
+  classes: string;
+}
+
+interface TemplateElement extends HTMLElement {
+  getTemplate(): Promise<ResultTemplate<DocumentFragment> | null>;
+}
+
+interface ResultListCommonOptions {
+  bindings: Bindings;
+  templateElements: NodeListOf<TemplateElement>;
+  fieldsToInclude?: string;
+  includeDefaultTemplate?: boolean;
+  onReady(): void;
+  onError(): void;
+}
+
+export class ResultListCommon {
+  private bindings: Bindings;
+  private render?: ResultRenderingFunction;
+  private updateBreakpoints?: (host: HTMLElement) => void;
+
+  public resultTemplatesManager!: ResultTemplatesManager<TemplateContent>;
+  public resultListControllerProps?: ResultListProps;
+
+  constructor(opts: ResultListCommonOptions) {
+    this.bindings = opts.bindings;
+    this.updateBreakpoints = once((host: HTMLElement) => {
+      updateBreakpoints(host);
+    });
+
+    if (opts.fieldsToInclude) {
+      this.resultListControllerProps = {
+        options: {
+          fieldsToInclude: this.determineAllFieldsToInclude(
+            opts.fieldsToInclude
+          ),
+        },
+      };
+    }
+
+    this.registerResultTemplates(
```

Comment:
since `registerResultTemplates` is async and public, the class which initializes `ResultListCommon` could call and await `registerResultTemplates` instead of passing `onReady` and `onError`